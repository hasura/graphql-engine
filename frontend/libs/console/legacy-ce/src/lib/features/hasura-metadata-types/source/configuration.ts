export type FromEnv = { from_env: string };
type ValidJson = Record<string, any>;

export interface PostgresConfiguration {
  connection_info: {
    database_url:
      | string
      | FromEnv
      | {
          username: string;
          password?: string;
          database: string;
          host: string;
          port: string;
        };
    pool_settings?: {
      max_connections?: number;
      idle_timeout?: number;
      retries?: number;
      pool_timeout?: number;
      connection_lifetime?: number;
    };
    use_prepared_statements?: boolean;
    /**
     * The transaction isolation level in which the queries made to the source will be run with (default: read-committed).
     */
    isolation_level?: 'read-committed' | 'repeatable-read' | 'serializable';
    ssl_configuration?: {
      sslmode: string;
      sslrootcert: FromEnv;
      sslcert: FromEnv;
      sslkey: FromEnv;
      sslpassword: FromEnv;
    };
  };
  /**
   * Optional list of read replica configuration (supported only in cloud/enterprise versions)
   */
  read_replicas?: PostgresConfiguration['connection_info'][];
  /**
   * Name of the schema where the graphql-engine will install database extensions (default: public)
   */
  extensions_schema?: any;
}

export interface MssqlConfiguration {
  connection_info: {
    connection_string: string | FromEnv;
    pool_settings?: {
      max_connections?: number;
      idle_timeout?: number;
    };
  };
  read_replicas?: MssqlConfiguration['connection_info'][];
}

export interface BigQueryConfiguration {
  service_account: string | ValidJson | FromEnv;
  project_id: string | FromEnv;
  datasets: string[] | FromEnv;
}

export interface CitusConfiguration extends PostgresConfiguration {}
