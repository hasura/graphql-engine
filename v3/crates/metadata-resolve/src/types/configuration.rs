/// Configuration for the metadata-resolve step.
///
/// Deserialization is only intended to be used for testing and is not reliable.
#[derive(Debug, Clone, Default, serde::Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
pub struct Configuration {
    pub allow_unknown_subgraphs: bool,
    pub unstable_features: UnstableFeatures,
}

/// internal feature flags used in metadata resolve steps
///
/// Deserialization is only intended to be used for testing and is not reliable.
#[derive(Debug, Clone, Copy, Default, serde::Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
#[allow(clippy::struct_excessive_bools)]
pub struct UnstableFeatures {
    pub enable_order_by_expressions: bool,
    pub enable_ndc_v02_support: bool,
    pub enable_subscriptions: bool,
    pub enable_jsonapi: bool,
}
