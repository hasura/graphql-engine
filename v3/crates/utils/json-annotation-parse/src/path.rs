use crate::types::Value;
use jsonpath::{JSONPath, JSONPathElement};

#[derive(Debug)]
pub enum WalkError<Ann> {
    PathValueMismatch(JSONPathElement, Value<Ann>),
    MissingIndex(usize, Value<Ann>),
    MissingKey(String, Value<Ann>),
}

// reverse the path then walk it
pub fn walk<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    let mut reversed_path = JSONPath(path.0.iter().rev().cloned().collect());
    walk_inner(value, &mut reversed_path)
}

fn walk_inner<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    match path.0.pop() {
        Some(tail) => {
            match (value, &tail) {
                (Value::Array(_, items), JSONPathElement::Index(i)) => {
                    match items.get(*i) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingIndex(*i, value.clone())),
                    }
                }
                (Value::Object(_, items), JSONPathElement::Key(s)) => {
                    match items.get(s) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingKey(s.to_string(), value.clone())),
                    }
                }
                _ => Err(WalkError::PathValueMismatch(tail.clone(), value.clone())),
            }
        }
        None => {
            // we are here, return outer annotation
            match value {
                Value::Bool(ann, _) => Ok(ann),
                Value::Null(ann) => Ok(ann),
                Value::Number(ann, _) => Ok(ann),
                Value::String(ann, _) => Ok(ann),
                Value::Array(ann, _) => Ok(ann),
                Value::Object(ann, _) => Ok(ann),
            }
        }
    }
}
