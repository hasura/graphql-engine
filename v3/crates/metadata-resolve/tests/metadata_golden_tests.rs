//! Tests that attempt to resolve different metadata files and assert that they parse successfully
//! or fail in the expected way.

use std::fs;
use std::path::Path;

use metadata_resolve::configuration;

#[test]
fn test_passing_metadata() {
    insta::glob!("passing/**/metadata.json", |path| {
        let directory = path.parent().unwrap();
        insta::with_settings!({
            snapshot_path => directory,
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let configuration = read_test_configuration(directory)
                .unwrap_or_else(|error| panic!("{}: Could not read configuration: {error}",directory.display()));

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("{}: Could not read file {path:?}: {error}", directory.display()));

            let metadata_json_value = serde_json::from_str(&metadata_json_text)
                .unwrap_or_else(|error| panic!("{}: Could not parse JSON: {error}",directory.display()));

            let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)
                .unwrap_or_else(|error| panic!("{}: Could not deserialize metadata: {error}", directory.display()));

            let resolved = metadata_resolve::resolve(metadata, &configuration)
                .unwrap_or_else(|error| panic!("{}: Could not resolve metadata: {error}",directory.display()));

            insta::assert_debug_snapshot!("resolved", resolved);
        });
    });
}

#[test]
fn test_failing_metadata() {
    insta::glob!("failing/**/metadata.json", |path| {
        let directory = path.parent().unwrap();
        insta::with_settings!({
            snapshot_path => directory,
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let configuration = read_test_configuration(directory)
                .unwrap_or_else(|error| panic!("{}: Could not read configuration: {error}", directory.display()));

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("{}: Could not read file {path:?}: {error}", directory.display()));

            match serde_json::from_str(&metadata_json_text) {
                Ok(metadata_json_value) => {
                    match open_dds::traits::OpenDd::deserialize(metadata_json_value) {
                        Ok(metadata) => {
                            match metadata_resolve::resolve(metadata, &configuration) {
                                Ok(_) => {
                                    panic!("{}: Unexpected success when resolving {path:?}.", directory.display());
                                }
                                Err(msg) => {
                                    insta::assert_snapshot!("resolve_error", msg);
                                }
                            }
                        }
                        Err(msg) => {
                            insta::assert_snapshot!("deserialize_error", msg);
                        }
                    };
                }

                Err(msg) => {
                    insta::assert_snapshot!("parse_error", msg);
                }
            };
        });
    });
}

fn read_test_configuration(
    directory: &Path,
) -> Result<configuration::Configuration, Box<dyn std::error::Error>> {
    let unstable_features = configuration::UnstableFeatures {
        enable_order_by_expressions: false,
        enable_ndc_v02_support: false,
        enable_subscriptions: false,
        enable_jsonapi: false,
    };

    let configuration_path = directory.join("configuration.json");
    if configuration_path.exists() {
        let reader = fs::File::open(configuration_path)?;
        let configuration = serde_json::from_reader(reader)?;
        Ok(configuration::Configuration {
            unstable_features,
            ..configuration
        })
    } else {
        Ok(configuration::Configuration {
            allow_unknown_subgraphs: false,
            unstable_features,
        })
    }
}
