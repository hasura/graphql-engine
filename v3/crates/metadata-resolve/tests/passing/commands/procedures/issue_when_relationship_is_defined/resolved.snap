---
source: crates/metadata-resolve/tests/metadata_golden_tests.rs
expression: resolved
input_file: crates/metadata-resolve/tests/passing/commands/procedures/issue_when_relationship_is_defined/metadata.json
---
(
    Metadata {
        object_types: {
            Qualified {
                subgraph: SubgraphName(
                    "default",
                ),
                name: CustomTypeName(
                    Identifier(
                        "MyObject",
                    ),
                ),
            }: ObjectTypeWithRelationships {
                object_type: ObjectTypeRepresentation {
                    fields: {
                        FieldName(
                            Identifier(
                                "id",
                            ),
                        ): FieldDefinition {
                            field_type: QualifiedTypeReference {
                                underlying_type: Named(
                                    Inbuilt(
                                        Int,
                                    ),
                                ),
                                nullable: false,
                            },
                            description: None,
                            deprecated: None,
                            field_arguments: {},
                        },
                        FieldName(
                            Identifier(
                                "name",
                            ),
                        ): FieldDefinition {
                            field_type: QualifiedTypeReference {
                                underlying_type: Named(
                                    Inbuilt(
                                        String,
                                    ),
                                ),
                                nullable: false,
                            },
                            description: None,
                            deprecated: None,
                            field_arguments: {},
                        },
                    },
                    global_id_fields: [],
                    apollo_federation_config: None,
                    graphql_output_type_name: Some(
                        TypeName(
                            Name(
                                "MyObject",
                            ),
                        ),
                    ),
                    graphql_input_type_name: None,
                    description: None,
                },
                type_output_permissions: {},
                type_input_permissions: {},
                relationship_fields: {
                    RelationshipName(
                        Identifier(
                            "procedureResult",
                        ),
                    ): RelationshipField {
                        field_name: Name(
                            "procedureResult",
                        ),
                        relationship_name: RelationshipName(
                            Identifier(
                                "procedureResult",
                            ),
                        ),
                        source: Qualified {
                            subgraph: SubgraphName(
                                "default",
                            ),
                            name: CustomTypeName(
                                Identifier(
                                    "MyObject",
                                ),
                            ),
                        },
                        target: Command(
                            CommandRelationshipTarget {
                                command_name: Qualified {
                                    subgraph: SubgraphName(
                                        "default",
                                    ),
                                    name: CommandName(
                                        Identifier(
                                            "procedure_with_args",
                                        ),
                                    ),
                                },
                                target_type: QualifiedTypeReference {
                                    underlying_type: Named(
                                        Inbuilt(
                                            String,
                                        ),
                                    ),
                                    nullable: true,
                                },
                                mappings: [
                                    RelationshipCommandMapping {
                                        source_field: RelationshipFieldAccess {
                                            field_name: FieldName(
                                                Identifier(
                                                    "id",
                                                ),
                                            ),
                                        },
                                        argument_name: ArgumentName(
                                            Identifier(
                                                "argB",
                                            ),
                                        ),
                                    },
                                ],
                            },
                        ),
                        target_capabilities: Some(
                            RelationshipCapabilities {
                                foreach: (),
                                supports_relationships: Some(
                                    DataConnectorRelationshipCapabilities {
                                        supports_relation_comparisons: true,
                                        supports_nested_relationships: Some(
                                            DataConnectorNestedRelationshipCapabilities {
                                                supports_nested_array_selection: true,
                                                supports_nested_in_filtering: false,
                                                supports_nested_in_ordering: false,
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                        description: None,
                        deprecated: None,
                    },
                },
                type_mappings: DataConnectorTypeMappingsForObject {
                    mappings: {},
                },
            },
        },
        scalar_types: {
            Qualified {
                subgraph: SubgraphName(
                    "default",
                ),
                name: CustomTypeName(
                    Identifier(
                        "Headers",
                    ),
                ),
            }: ScalarTypeRepresentation {
                graphql_type_name: Some(
                    TypeName(
                        Name(
                            "Headers",
                        ),
                    ),
                ),
                description: Some(
                    "Headers map",
                ),
                representations: {
                    Qualified {
                        subgraph: SubgraphName(
                            "default",
                        ),
                        name: DataConnectorName(
                            Identifier(
                                "myconnector",
                            ),
                        ),
                    }: JSON,
                },
            },
        },
        models: {},
        commands: {
            Qualified {
                subgraph: SubgraphName(
                    "default",
                ),
                name: CommandName(
                    Identifier(
                        "procedure_with_args",
                    ),
                ),
            }: CommandWithPermissions {
                command: Command {
                    name: Qualified {
                        subgraph: SubgraphName(
                            "default",
                        ),
                        name: CommandName(
                            Identifier(
                                "procedure_with_args",
                            ),
                        ),
                    },
                    output_type: QualifiedTypeReference {
                        underlying_type: Named(
                            Inbuilt(
                                String,
                            ),
                        ),
                        nullable: true,
                    },
                    arguments: {
                        ArgumentName(
                            Identifier(
                                "argA",
                            ),
                        ): ArgumentInfo {
                            argument_type: QualifiedTypeReference {
                                underlying_type: Named(
                                    Inbuilt(
                                        String,
                                    ),
                                ),
                                nullable: false,
                            },
                            description: None,
                            argument_kind: Other,
                            path: JSONPath(
                                [
                                    Key(
                                        "subgraphs",
                                    ),
                                    Index(
                                        0,
                                    ),
                                    Key(
                                        "objects",
                                    ),
                                    Index(
                                        2,
                                    ),
                                    Key(
                                        "definition",
                                    ),
                                    Key(
                                        "arguments",
                                    ),
                                    Index(
                                        0,
                                    ),
                                    Key(
                                        "name",
                                    ),
                                ],
                            ),
                        },
                        ArgumentName(
                            Identifier(
                                "argB",
                            ),
                        ): ArgumentInfo {
                            argument_type: QualifiedTypeReference {
                                underlying_type: Named(
                                    Inbuilt(
                                        Int,
                                    ),
                                ),
                                nullable: false,
                            },
                            description: None,
                            argument_kind: Other,
                            path: JSONPath(
                                [
                                    Key(
                                        "subgraphs",
                                    ),
                                    Index(
                                        0,
                                    ),
                                    Key(
                                        "objects",
                                    ),
                                    Index(
                                        2,
                                    ),
                                    Key(
                                        "definition",
                                    ),
                                    Key(
                                        "arguments",
                                    ),
                                    Index(
                                        1,
                                    ),
                                    Key(
                                        "name",
                                    ),
                                ],
                            ),
                        },
                    },
                    graphql_api: None,
                    source: Some(
                        CommandSource {
                            data_connector: DataConnectorLink {
                                name: Qualified {
                                    subgraph: SubgraphName(
                                        "default",
                                    ),
                                    name: DataConnectorName(
                                        Identifier(
                                            "myconnector",
                                        ),
                                    ),
                                },
                                url: ReadWriteUrls(
                                    ResolvedReadWriteUrls {
                                        read: SerializableUrl(
                                            Url {
                                                scheme: "http",
                                                cannot_be_a_base: false,
                                                username: "",
                                                password: None,
                                                host: Some(
                                                    Domain(
                                                        "local-dev.hasura.me",
                                                    ),
                                                ),
                                                port: Some(
                                                    8080,
                                                ),
                                                path: "/",
                                                query: None,
                                                fragment: None,
                                            },
                                        ),
                                        write: SerializableUrl(
                                            Url {
                                                scheme: "http",
                                                cannot_be_a_base: false,
                                                username: "",
                                                password: None,
                                                host: Some(
                                                    Domain(
                                                        "local-dev.hasura.me",
                                                    ),
                                                ),
                                                port: Some(
                                                    8080,
                                                ),
                                                path: "/",
                                                query: None,
                                                fragment: None,
                                            },
                                        ),
                                    },
                                ),
                                headers: SerializableHeaderMap(
                                    {},
                                ),
                                response_config: None,
                                capabilities: DataConnectorCapabilities {
                                    supported_ndc_version: V01,
                                    supports_explaining_queries: true,
                                    supports_explaining_mutations: true,
                                    supports_nested_object_filtering: false,
                                    supports_nested_object_ordering: false,
                                    supports_nested_object_array_filtering: false,
                                    supports_nested_scalar_array_filtering: false,
                                    supports_aggregates: Some(
                                        DataConnectorAggregateCapabilities {
                                            supports_nested_object_aggregations: false,
                                            aggregate_count_scalar_type: None,
                                            supports_grouping: None,
                                        },
                                    ),
                                    supports_query_variables: true,
                                    supports_relationships: Some(
                                        DataConnectorRelationshipCapabilities {
                                            supports_relation_comparisons: true,
                                            supports_nested_relationships: Some(
                                                DataConnectorNestedRelationshipCapabilities {
                                                    supports_nested_array_selection: true,
                                                    supports_nested_in_filtering: false,
                                                    supports_nested_in_ordering: false,
                                                },
                                            ),
                                        },
                                    ),
                                    supports_relational_queries: None,
                                },
                            },
                            source: Procedure(
                                ProcedureName(
                                    "procedure_with_args",
                                ),
                            ),
                            ndc_type_opendd_type_same: true,
                            type_mappings: {},
                            argument_mappings: {
                                ArgumentName(
                                    Identifier(
                                        "argA",
                                    ),
                                ): DataConnectorArgumentName(
                                    "arg1",
                                ),
                                ArgumentName(
                                    Identifier(
                                        "argB",
                                    ),
                                ): DataConnectorArgumentName(
                                    "arg2",
                                ),
                            },
                            data_connector_link_argument_presets: {
                                DataConnectorArgumentName(
                                    "headers",
                                ): ArgumentPresetValue {
                                    http_headers: HttpHeadersPreset {
                                        forward: [
                                            SerializableHeaderName(
                                                "x-myheader",
                                            ),
                                        ],
                                        additional: {
                                            SerializableHeaderName(
                                                "x-whatever",
                                            ): Literal(
                                                String("wow"),
                                            ),
                                        },
                                    },
                                },
                            },
                            source_arguments: {
                                DataConnectorArgumentName(
                                    "arg1",
                                ): Named {
                                    name: TypeName(
                                        "String",
                                    ),
                                },
                                DataConnectorArgumentName(
                                    "arg2",
                                ): Named {
                                    name: TypeName(
                                        "Int",
                                    ),
                                },
                                DataConnectorArgumentName(
                                    "headers",
                                ): Named {
                                    name: TypeName(
                                        "Headers",
                                    ),
                                },
                            },
                        },
                    ),
                    description: None,
                },
                permissions: {},
            },
        },
        boolean_expression_types: BooleanExpressionTypes {
            objects: {},
            scalars: {},
            object_aggregates: {},
            scalar_aggregates: {},
        },
        order_by_expressions: OrderByExpressions {
            objects: {},
            scalars: {},
        },
        aggregate_expressions: {},
        graphql_config: GlobalGraphqlConfig {
            query_root_type_name: TypeName(
                Name(
                    "Query",
                ),
            ),
            mutation_root_type_name: TypeName(
                Name(
                    "Mutation",
                ),
            ),
            subscription_root_type_name: None,
            order_by_input: Some(
                OrderByInputGraphqlConfig {
                    asc_direction_field_value: Name(
                        "Asc",
                    ),
                    desc_direction_field_value: Name(
                        "Desc",
                    ),
                    enum_type_name: TypeName(
                        Name(
                            "order_by",
                        ),
                    ),
                },
            ),
            enable_apollo_federation_fields: false,
            bypass_relation_comparisons_ndc_capability: false,
            propagate_boolean_expression_deprecation_status: false,
            multiple_order_by_input_object_fields: Allow,
        },
        plugin_configs: LifecyclePluginConfigs {
            pre_parse_plugins: [],
            pre_response_plugins: [],
            pre_route_plugins: [],
        },
        roles: {},
        runtime_flags: RuntimeFlags(
            {},
        ),
    },
    [
        ObjectRelationshipsIssue(
            ProcedureCommandRelationshipsNotSupported {
                type_name: Qualified {
                    subgraph: SubgraphName(
                        "default",
                    ),
                    name: CustomTypeName(
                        Identifier(
                            "MyObject",
                        ),
                    ),
                },
                relationship_name: RelationshipName(
                    Identifier(
                        "procedureResult",
                    ),
                ),
                command_name: Qualified {
                    subgraph: SubgraphName(
                        "default",
                    ),
                    name: CommandName(
                        Identifier(
                            "procedure_with_args",
                        ),
                    ),
                },
                procedure_name: ProcedureName(
                    "procedure_with_args",
                ),
            },
        ),
    ],
)
